src/server_state.cpp

```diff
@@ -60,25 +60,91 @@ static void log_adopt_error(const wsrep::transaction& transaction)
                          << "which may need to be removed manually.";
 }
 
-static int apply_fragment(wsrep::high_priority_service& high_priority_service,
-                          wsrep::high_priority_service& streaming_applier,
+// resolve which of the two errors return to caller
+static inline int resolve_return_error(bool const vote,
+                                       int  const vote_err,
+                                       int  const apply_err)
+{
+    if (vote) return vote_err;
+    return vote_err != 0 ? vote_err : apply_err;
+}
+
+static void
+discard_streaming_applier(wsrep::server_state& server_state,
+                          wsrep::high_priority_service* streaming_applier,
+                          const wsrep::ws_meta& ws_meta)
+{
+    server_state.stop_streaming_applier(
+        ws_meta.server_id(), ws_meta.transaction_id());
+    server_state.server_service().release_high_priority_service(
+        streaming_applier);
+}
+
+static int apply_fragment(wsrep::server_state& server_state,
+                          wsrep::high_priority_service& high_priority_service,
+                          wsrep::high_priority_service* streaming_applier,
                           const wsrep::ws_handle& ws_handle,
                           const wsrep::ws_meta& ws_meta,
                           const wsrep::const_buffer& data)
 {
-    int ret;
+    int ret(0);
+    int apply_err;
```

Comment:
`apply_err` could be moved inside block below.

Reply:
It will be needed outside the block. I don't see how to avoid it.